### DISCUSSION TITLE
generateStaticParams() not working for me!!!!

### DISCUSSION THREAD

**Discussion by sably** [vercel/next.js#50254] [App Router] [answered]
in the following code:

```javascript
//  /posts/[id].js

import Layout from '../../components/layout';

export function generateStaticParams() {
    return [{ id: '1' }, { id: '2' }, { id: '3' }];
}

export default function Page({params}) {
    console.log(params)
    return (
        <Layout>
               <div>test</div>
        </Layout>
    );
}
``` 

when I navigate to /post/1 console show undefined for params!!
can u help me?

**Comment by icyJoseph** [comment]
Hi,

why do you have `next/head`? Makes me believe that you are using the `pages` directory, and `generateStaticParams` is an `app` directory function

Also your page is called `/posts/[id].js`, which is a pages directory pattern, in `app` directory it should be `/posts/[id]/page.js`

**Reply by sably** [reply]
u mean we cant use [id].js?

**Reply by icyJoseph** [reply]
It has to be `/posts/[id]/page.js`, please refer to the file conventions: https://nextjs.org/docs/app/api-reference/file-conventions/page

**Reply by sably** [reply]
see this [https://nextjs.org/docs/app/api-reference/functions/generate-static-params#single-dynamic-segment](url)

**Reply by icyJoseph** [reply]
Yes, it is showing you the same thing I am saying, you need to end it with `/posts/[id]/page.js`

![Screenshot 2023-05-24 at 09 26 39](https://github.com/vercel/next.js/assets/21013447/9a3b565e-f37e-4e9e-8530-df20cacc5e8a)

Look at the file name above the code.

**Reply by sably** [reply]
change code to javascript not typescript

**Reply by sably** [reply]
![Screenshot 2023-05-24 110056](https://github.com/vercel/next.js/assets/25221341/440addca-70b5-4eb1-877a-52de94188392)

**Reply by icyJoseph** [reply]
I see, that's an unfortunate error in the documentation, it should be fixed/reported.

**Reply by icyJoseph** [reply]
Did it work for you when using `app/posts/[id]/page.js` though? I've sent a feedback message on that page, hopefully it is fixed soon. Most likely not the first/last to run into this issue.

**Reply by sably** [reply]
nop, when i changed dir to  /posts/[id]/page.js see what happend:
![ss](https://github.com/vercel/next.js/assets/25221341/4b1a06b7-1f08-4d90-bcf0-038908272f89)

**Reply by sably** [reply]
and i have a 404.js page nested to posts directory, not inside that

**Reply by sably** [reply]
my dir:
![bbbbbb](https://github.com/vercel/next.js/assets/25221341/b1f3e477-13d1-4b19-a594-41d72d6e57de)

**Reply by icyJoseph** [reply]
mmm it works for me... 

![Screenshot 2023-05-24 at 09 50 04](https://github.com/vercel/next.js/assets/21013447/cd8f826e-f674-45ee-803c-f6d97bf50009)

Could you share your file tree?

**Reply by icyJoseph** [reply, marked as answer]
Right, I asked you if you are using `pages` or `app`

You are using `pages` directory. That's an entirely different tech stack.

1. Change your file to `pages/posts/[id].js` - yes, roll back
2. use `getStaticPaths` - https://nextjs.org/docs/pages/building-your-application/data-fetching/get-static-paths
3. use `useRouter` and from there the `query` property, and from the `query`, the `id`.

The `generateStaticParams` only works in `app` directory.

**Reply by sably** [reply]
ty so much 
and damn to nextjs ducumantaion

**Reply by icyJoseph** [reply]
No worries, you can switch between pages/app documentation  

![Screenshot 2023-05-24 at 10 12 22](https://github.com/vercel/next.js/assets/21013447/41eb2c70-63f1-4cf9-b41e-ca09f6671272)

Have a good day!
